scripts/migration/lib.sh

```diff
@@ -369,6 +411,112 @@ append_trap() {
+}
+
+unlock_upgrade() {
+  kubectl_do "${1}" delete configmap --namespace kube-system apps-upgrade &>/dev/null ||
+    log_warn "Could not unlock upgrade or upgrade not locked"
+}
+
+# Get currently prepared version
+get_prepared_version() {
+  kubectl_do "${1}" get cm --namespace kube-system apps-upgrade --output=jsonpath --template="{.data.version}"
+}
+
+check_prepared_version() {
+  local prepared_version
+  prepared_version="$(get_prepared_version "${1}" || true)"
+  if [ -z "${prepared_version}" ]; then
+    log_fatal "'prepare' step does not appear to have been run, do so first"
+  fi
+
+  if [[ "${prepared_version}" != "${CK8S_TARGET_VERSION}" ]]; then
+    log_fatal "'prepare' step in ${1} appears to have been run for version ${prepared_version}, not ${CK8S_TARGET_VERSION}"
+  fi
+}
+
+# Usage: record_upgrade_prepare_done sc|wc
+# Records that upgrade has been prepared by storing a timestamp in the config and in the cluster along with the target
+# version.
+record_upgrade_prepare_done() {
+  local apps_config_timestamp
+  apps_config_timestamp="$(date -uIs)"
+
+  # This ConfigMap should only exist while doing an upgrade.
+  # Abort if it already exists
+  log_info "Locking cluster ${1} for upgrade"
+  if kubectl_do "${1}" create configmap --dry-run=client --output=yaml \
+    --namespace kube-system apps-upgrade \
+    --from-literal "version=${CK8S_TARGET_VERSION}" \
+    --from-literal "timestamp=${apps_config_timestamp}" |
+    yq '.metadata.labels["app.kubernetes.io/managed-by"] = "apps-upgrade"' - |
+    kubectl_do "${1}" create --filename -; then
+
+    ts="${apps_config_timestamp}" \
+      yq_add common '.global.ck8sConfigSerial' 'strenv(ts)'
+    log_info "Cluster ${1} locked for upgrade"
+    return 0
+  else
+    log_warn "prepare already started in ${1} ('ck8s upgrade ${1} unlock' to try again)"
+  fi
+}
+
+# Usage: ensure_upgrade_prepared sc|wc
+# Ensures that the timestamp and version recorded in the config matches that recorded in the cluster.
+ensure_upgrade_prepared() {
+  local apps_upgrade
+  local apps_version
+  local apps_cluster_timestamp
+  local apps_config_timestamp
+
+  apps_upgrade="$(kubectl_do "${1}" get --namespace kube-system cm apps-upgrade --output=yaml)"
+  apps_version="$(yq '.data.version' <<<"${apps_upgrade}")"
+
+  if [[ "${apps_version}" != "${CK8S_TARGET_VERSION}" ]] >/dev/null; then
+    log_fatal "version mismatch, upgrading to ${CK8S_TARGET_VERSION} but cluster ${1} was prepared for ${apps_version}"
+  fi
+
+  apps_config_timestamp="$(yq '.data.timestamp' <<<"${apps_upgrade}")"
+  apps_cluster_timestamp="$(yq_dig common '.global.ck8sConfigSerial')"
+  if [[ "${apps_config_timestamp}" != "${apps_cluster_timestamp}" ]]; then
+    log_fatal "Config timestamp mismatch, ${apps_cluster_timestamp} in ${1} but ${apps_config_timestamp} in config"
+  fi
+}
+
+# Usage: record_upgrade_apply_step sc|wc step-description
+record_upgrade_apply_step() {
+  # Was this needed?
```

Comment:
Was it? :smile:

Reply:
_Need_ is such a strong word. I was wondering if there was consensus on whether to keep or remove this. Easy enough to remove later anyway. Wrote a more descriptive comment.